File: 678.cpp
struct Solution {
    bool checkValidString(string &s) {
        int depMin = 0, depMax = 0;
        for (char &c : s){
            switch (c){
                case '(':
                    ++depMin;
                    ++depMax;
                    break;
                case ')':
                    depMin = max(0, depMin - 1);
                    --depMax;
                    break;
                case '*':
                    depMin = max(0, depMin - 1);
                    ++depMax;
            }
            if (depMax < 0){
                return false;
            }
        }
        return depMin == 0;
    }
};

File: sol/678.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        depMin, depMax = 0, 0
        for c in s:
            if c == '(':
                depMin += 1
                depMax += 1
            elif c == ')':
                depMin -= 1
                depMax -= 1
            elif c == '*':
                depMin -= 1
                depMax += 1
            if depMax < 0:
                return False
            depMin = max(0, depMin)
        return depMin == 0
